### PR TITLE
set_dir expanding "~"

### LINKED ISSUE #69761: torch.hub.set_dir() not expanding ~
### 🐛 Describe the bug

```python
import torch

torch.hub.set_dir("~/example_dir/")
torch.hub.get_dir()  # Returns the string "~/example_dir/"
```

When the string from get_dir() is used ([_get_cache_or_reload](https://github.com/pytorch/pytorch/blob/master/torch/hub.py#L164) or [load_state_dict_from_url](https://github.com/pytorch/pytorch/blob/master/torch/hub.py#L567)), then it will create a directory called literally "~/example_dir/" in the current directory (I have verified this locally).

```python
os.makedirs("~/example_dir/")  # Will create a directory in the current directory called ./~/example_dir
```

Proposed fix: 
torch.hub.set_dir() should expand tildes using os.path.expanduser, like in [get_torch_home()](https://github.com/pytorch/pytorch/blob/master/torch/hub.py#L105)

```python
def set_dir(d):
    r"""
    Optionally set the Torch Hub directory used to save downloaded models & weights.
    Args:
        d (string): path to a local folder to save downloaded models & weights.
    """
    global _hub_dir
    _hub_dir = os.path.expand_dirs(d)
```

I'll send in a PR with the above so long, but don't have the storage on my laptop to download + compile + test pytorch locally :/ 

Thanks!

### Versions

Collecting environment information...
PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 10.15.7 (x86_64)
GCC version: Could not collect
Clang version: 11.0.3 (clang-1103.0.32.59)
CMake version: version 3.19.4
Libc version: N/A

Python version: 3.7.9 (default, Aug 31 2020, 07:22:35)  [Clang 10.0.0 ] (64-bit runtime)
Python platform: Darwin-19.6.0-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] torch==1.7.1
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      233
[conda] mkl-service               2.3.0            py37h9ed2024_0
[conda] mkl_fft                   1.3.0            py37ha059aab_0
[conda] mkl_random                1.1.1            py37h959d312_0
[conda] numpy                     1.19.1           py37h3b9f5b6_0    anaconda
[conda] numpy-base                1.19.1           py37hcfb5961_0
[conda] pytorch                   1.7.1                   py3.7_0    pytorch

cc @nairbv @NicolasHug @vmoens @jdsgomes

### PR DESCRIPTION
**Pull request by loodvn**
Fixes #69761.

Small change to torch.hub.set_dir() (<10 LOC).

It seems that before the code was split into `set_dir()` and `_get_torch_home `, an [earlier version](https://github.com/pytorch/pytorch/blame/5164622ba462fe07fc9f2325fccf7f85aecb3ec8/torch/hub.py#L111) of hub.py had a os.path.expanduser check. 

Currently, [_get_torch_home](https://github.com/pytorch/pytorch/blob/master/torch/hub.py#L104) retained the os.path.expanduser check, but `set_dir()` didn't have one. This PR fixes that (I hope).

(As I mentioned in the issue, I can't run the tests on my laptop yet because of storage space :/ But I did include a test.)

### CODE REVIEW

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
Thanks for the quick fix. @loodvn please sign the CLA and your change is good to go

**Comment by loodvn** [comment]
@malfet Thanks! I signed the CLA btw, wasn’t sure if I should ping you afterwards or not 👍 (and thanks @NicolasHug for waiting, I just saw that 🙂)

**Comment by loodvn** [comment]
Hi there, could this still be merged? It's a one-line change + a test.

File: test/test_utils.py [hunk 1]
```diff
@@ -689,6 +689,14 @@ def test_hub_dir(self):
             torch.hub.set_dir(dirname)
             self.assertEqual(torch.hub.get_dir(), dirname)
 
+    def test_hub_dir_expand(self):
+        with tempfile.TemporaryDirectory('cur_dir') as dirname:
```
Comment (NicolasHug) [inline test/test_utils.py:693]: @loodvn Thanks for the PR, and for your patience. The PR looks good, but I'm wondering why we need a tempdir here? I think it comes from the previous test above, but we don't seem to use it in this test.

**Review by NicolasHug — APPROVED** [review]

**Review by malfet — APPROVED** [review]

**Comment by loodvn** [comment]
Thanks!

Label suggestions:
`release notes: hub`
`topic: bug fixes`

**Comment by malfet** [comment]
@loodvn applied, but you could have asked merged bot to do it for you. (note to self: need to update the message)

**Comment by NicolasHug** [comment]
Thanks a lot for the pr @loodvn !
